tolstoy/src/tx_client.rs

```diff
@@ -10,48 +10,61 @@
 
 use std::collections::BTreeMap;
 use std::collections::btree_map::Entry;
+use std::collections::HashMap;
 
 use rusqlite;
 
 use errors::{
     Result,
-    ErrorKind
-};
-
-use mentat_db::types::{
-    Entid
+    ErrorKind,
 };
 
 use mentat_db::{
     entids,
-    TypedSQLValue
+    TypedSQLValue,
 };
 
 use mentat_core::{
-    TypedValue
+    TypedValue,
+    Entid,
 };
 
 #[derive(Debug)]
 pub struct TxPart {
     pub e: Entid,
-    pub a: i64,
+    pub a: Entid,
     pub v: TypedValue,
-    pub added: i32
+    pub added: bool,
 }
 
+// Notes on 'parts' representation:
+// Currently it's suitable for uses which necessitate pulling in the entire tx into memory,
+// and don't require efficient querying/monitoring by attributes of parts.
+//
+// Example: streaming transactions to/from the server.
+//
+// In the future, consider:
+// - A structure that makes typical tx-listener questions — "does this transaction mention
+//   an attribute or entity I care about?" — efficient. That might be a trie, it might be a
+//   bunch of extra data structures (e.g., a set of mentioned attributes), or something else.
+//   With an unsorted Vec<TxPart>, looking for a mentioned attribute requires linear search of the entire vector.
+// - A structure that doesn't require pulling the entire tx into memory. This might be a cursor,
+//   a rowid range, or something else that's scoped to the lifetime of a particular database transaction,
+//   in order to preserve isolation.
 #[derive(Debug)]
 pub struct Tx {
     pub tx: Entid,
     pub tx_instant: TypedValue,
-    pub parts: Vec<TxPart>
+    pub parts: Vec<TxPart>,
 }
 
-struct RawTx {
+struct RawDatom {
     e: Entid,
-    a: i64,
-    v: TypedValue,
+    a: Entid,
+    v: TypedValue, // composite of 'v' and 'value_type_tag'
     tx: Entid,
-    added: i32
+    added: bool,
+    is_transaction: bool
```

Comment:
This is a bit of a smell, no?

The set of transactions is `SELECT DISTINCT tx FROM transactions`.

You're using the domain of `:db/txInstant` as a proxy for that, but (a) you can describe transactions that are not present in the local store (e.g., through excision), and (b) it's possible for calling code to do weird things like accidentally describe non-transactions with `:db/txInstant`.

What you mean here by `is_transaction` is "attribute is `:db/txInstant`", right?

There are plenty of other ways to figure out the set of txids and the timestamps of each without adding a column to the output, no? A comparison of `a == entids::DB_TX_INSTANT` is probably faster than using `is_transaction`, populating it from SQL, and adding a field to this structure.